server/data_soma/soma_adaptor.py

```diff
@@ -0,0 +1,338 @@
+            self.X_approximate_distribution = estimate_distribution.estimate_approximate_distribution(m)
+
+        return self.X_approximate_distribution
+
+    def get_shape(self):
+        """return the shape of the data matrix X"""
+        return self.data.X.data.shape()
+
+    def query_var_array(self, term_var):
+        return self.get_var_df()[term_var]
+
+    def query_obs_array(self, term_var):
+        return self.get_obs_df()[term_var]
+
+    def get_colors(self):
+        return convert_soma_category_colors_to_cxg_category_colors(self.data)
+
+    def get_obs_index(self):
+        name = self.server_config.single_dataset__obs_names
+        if name is None:
+            return self.original_obs_index
+        else:
+            return self.data.obs.df()[name]
+
+    def get_obs_columns(self):
+        return self.data.obs.df().columns
+
+    def get_obs_keys(self):
+        # return list of keys
+        return self.data.obs.keys()
+
+    def get_var_keys(self):
+        # return list of keys
+        return self.data.var.keys()
+
+    def get_obs_df(self):
+        # get obs dataframe in the usual format, indexed by integer indices, not obs_id
+        df = self.data.obs.df()
+        df = df.rename_axis(self.get_schema()["annotations"]["obs"]["index"]).reset_index()
+        return df
+
+    def get_var_df(self):
+        # get var dataframe in the usual format, indexed by integer indices, not var_id
+        df = self.data.var.df()
+        df = df.rename_axis(self.get_schema()["annotations"]["var"]["index"]).reset_index()
+        return df
+
+    def cleanup(self):
+        pass
+
+    def _create_schema(self):
+        self.schema = {
+            "dataframe": {
+                "nObs": self.cell_count,
+                "nVar": self.gene_count,
+                **get_schema_type_hint_of_array(self.data.X.data.df().to_numpy()),
+            },
+            "annotations": {
+                "obs": {"index": self.parameters.get("obs_names"), "columns": []},
+                "var": {"index": self.parameters.get("var_names"), "columns": []},
+            },
+            "layout": {"obs": []},
+        }
+        for ax in Axis:
+            curr_axis = None
+            if str(ax) == "obs":
+                curr_axis = self.get_obs_df()
+            elif str(ax) == "var":
+                curr_axis = self.get_var_df()
+            else:
+                raise RuntimeError("Invalid axis: %s." % str(ax))
+            for soma in curr_axis:
+                soma_schema = {"name": soma, "writable": False}
+                soma_schema.update(get_schema_type_hint_of_array(curr_axis[soma]))
+                self.schema["annotations"][ax]["columns"].append(soma_schema)
+
+        for layout in self.get_embedding_names():
+            layout_schema = {"name": layout, "type": "float32", "dims": [f"{layout}_0", f"{layout}_1"]}
+            self.schema["layout"]["obs"].append(layout_schema)
+
+    def get_schema(self):
```

Comment:
`_create_schema()` is called when `SomaAdaptor` is initialized, so that's not going to happen unless someone modifies the class. I also understood it from `anndata_adaptor.py` as getting the "current" schema, so if it hasn't been "created" yet, then it should return None (if someone has the ability to call `get_schema()`, they can also call `create_schema()`, so it would be odd if the getter was also the setter/creator).